configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables

alejandro-colomar opened the pull request:
Closes: <https://github.com/shadow-maint/shadow/issues/1229>
Reported-by: @zeha 
Cc: @thesamesam 

---

-  Queued after <https://github.com/shadow-maint/shadow/pull/1468>.

---

Is this okay?

---

Revisions:

<details>
<summary>v2</summary>

-  Remove definitions of $prefix and $exec_prefix.  [@thesamesam ]

```
$ git range-diff master gh/ep ep 
1:  8840d0e3 ! 1:  6f4e5d61 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
    @@ Metadata
      ## Commit message ##
         configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
     
    +    According to Sam, if we just remove the definitions, autotools will
    +    provide the right values.
    +
    +    Note that this changes the default from /usr to /usr/local, which is the
    +    right value according to the GNU conding standards (and also what the
    +    BSDs prefer, according to one conversation I had with Ingo Schwarze from
    +    OpenBSD).
    +
         Closes: <https://github.com/shadow-maint/shadow/issues/1229>
    +    Link: <https://www.gnu.org/software/autoconf/manual/autoconf-2.72/html_node/Default-Prefix.html>
         Reported-by: Chris Hofstaedtler <zeha@debian.org>
         Cc: Sam James <sam@gentoo.org>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
    @@ src/Makefile.am
      
     -ubindir = ${prefix}/bin
     -usbindir = ${prefix}/sbin
    -+prefix = /usr/local
    -+exec_prefix = ${prefix}
     +ubindir = ${exec_prefix}/bin
     +usbindir = ${exec_prefix}/sbin
      suidperms = 4755
```
</details>

<details>
<summary>v3</summary>

-  Replace `${}` by `$()` in surrounding code.  [@thesamesam ]

```
$ git range-diff master gh/ep ep 
1:  6f4e5d61 = 1:  6f4e5d61 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
-:  -------- > 2:  0fada16b src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3b</summary>

-  Rebase

```
$ git range-diff master..gh/ep shadow/master..ep 
1:  6f4e5d61 = 1:  983e3308 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  0fada16b = 2:  8d0eb980 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3c</summary>

-  Rebase

```
$ git rd
1:  983e3308 = 1:  3d09f4f9 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  8d0eb980 = 2:  51b0e9f9 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3d</summary>

-  Rebase

```
$ git range-diff db/master..gh/ep shadow/master..ep 
1:  3d09f4f9 = 1:  d19fdb29 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  51b0e9f9 = 2:  33142670 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3e</summary>

-  Rebase

```
$ git rd
1:  d19fdb29 = 1:  c499e815 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  33142670 = 2:  8738df50 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3f</summary>

-  Rebase

```
$ git rd
1:  c499e815 = 1:  905759e6 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  8738df50 = 2:  ba3c8f90 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3g</summary>

-  Rebase

```
$ git rd 
1:  905759e6 = 1:  657b01f1 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  ba3c8f90 = 2:  9dff75be src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3h</summary>

-  Rebase

```
$ git rd
1:  657b01f1 = 1:  ca0daff9 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  9dff75be = 2:  3f86cfbb src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3i</summary>

-  Rebase

```
$ git rd
1:  ca0daff9 = 1:  f793375e configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  3f86cfbb = 2:  cc9e79c7 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3j</summary>

-  Rebase

```
$ git rd
1:  f793375e = 1:  4309e187 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  cc9e79c7 = 2:  b28fb398 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3k</summary>

-  Rebase

```
$ git rd 
1:  4309e187 = 1:  e9fc9e1b configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  b28fb398 = 2:  7ab69930 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3l</summary>

-  Rebase

```
$ git rd 
1:  e9fc9e1b = 1:  0eeb7f4c configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  7ab69930 = 2:  e471ea0d src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3m</summary>

-  Rebase

```
$ git rd 
1:  0eeb7f4c = 1:  389a4058 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  e471ea0d = 2:  a7585cb7 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3n</summary>

-  Rebase

```
$ git rd 
1:  389a4058 = 1:  71ea3767 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  a7585cb7 = 2:  4f6598bb src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3o</summary>

-  Rebase

```
$ git rd 
1:  71ea3767 = 1:  57bd9370 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  4f6598bb = 2:  cf00ba24 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3p</summary>

-  Rebase

```
$ git rd 
1:  57bd9370 = 1:  89a30f10 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  cf00ba24 = 2:  0d718ce4 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v3q</summary>

-  Rebase

```
$ git rd 
1:  89a30f10 = 1:  800aaa5d configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  0d718ce4 = 2:  c3b6e0c3 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v4</summary>

-  Rebase

```
$ git rd 
1:  800aaa5d ! 1:  88acb4f1 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
    @@ configure.ac: AC_SUBST([LIBSUBID_ABI_MINOR], [libsubid_abi_minor])
      AC_SUBST([LIBSUBID_ABI], [libsubid_abi])
      
     -dnl Some hacks...
    --test "$prefix" = "NONE" && prefix="/usr"
    --test "$prefix" = "/usr" && exec_prefix=""
    +-test "x$prefix" = "xNONE" && prefix="/usr"
    +-test "X$prefix" = "X/usr" && exec_prefix=""
     -
      AC_USE_SYSTEM_EXTENSIONS
      
2:  c3b6e0c3 = 2:  fbcd57ed src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v4b</summary>

-  Rebase

```
$ git rd 
1:  88acb4f1 = 1:  aac8ef2f configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  fbcd57ed = 2:  e13042b0 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v4c</summary>

-  Rebase

```
$ git rd 
1:  aac8ef2f = 1:  7e76060c configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  e13042b0 = 2:  04adb6df src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v4d</summary>

-  Rebase

```
$ git rd 
1:  7e76060c = 1:  40f81475 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  04adb6df = 2:  f9deed24 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v4e</summary>

-  Rebase

```
$ git rd 
1:  40f81475 = 1:  bd86b89c configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  f9deed24 = 2:  c6816608 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v4f</summary>

-  Rebase

```
$ git rd 
1:  bd86b89c9 = 1:  b5bbddca8 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  c6816608b = 2:  64fade5f7 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v4g</summary>

-  Rebase

```
$ git rd 
1:  b5bbddca8 = 1:  d4743e4a1 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  64fade5f7 = 2:  8cacf6b31 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v4h</summary>

-  Rebase

```
$ git rd 
1:  d4743e4a1 = 1:  4ab4da022 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  8cacf6b31 = 2:  d7b978aae src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v5</summary>

-  Respect `$(sbindir)` in PASSWD_PROGRAM.  [@Karlson2k ]

```
$ git rd 
-:  --------- > 1:  5d4e20a93 configure.ac, lib/Makefile.am: Respect $(sbindir) in PASSWD_PROGRAM
1:  4ab4da022 = 2:  0ff5aa752 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  d7b978aae = 3:  921e7348c src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v5b</summary>

-  Escape double quotes, so that we have a string literal in the C code.

```
$ git rd 
1:  5d4e20a93 ! 1:  a884b9bd5 configure.ac, lib/Makefile.am: Respect $(sbindir) in PASSWD_PROGRAM
    @@ lib/Makefile.am
      
     -DEFS =
     +DEFS = \
    -+  -DPASSWD_PROGRAM="$(sbindir)/passwd"
    ++  -DPASSWD_PROGRAM='"$(sbindir)/passwd"'
      
      noinst_LTLIBRARIES = libshadow.la
      
2:  0ff5aa752 = 2:  86adba75d configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
3:  921e7348c = 3:  fc101c6c1 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v6</summary>

-  Rebase after #1468.

```
$ git range-diff master..gh/ep PASSWD_PROGRAM..ep 
1:  a884b9bd5 < -:  --------- configure.ac, lib/Makefile.am: Respect $(sbindir) in PASSWD_PROGRAM
2:  86adba75d = 1:  64bfdfd55 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
3:  fc101c6c1 = 2:  092b283d7 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v6b</summary>

-  Rebase

```
$ git range-diff 64bfdfd55a83^..gh/ep PASSWD_PROGRAM..ep 
1:  64bfdfd55 = 1:  e3239be5a configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  092b283d7 = 2:  b33540fc7 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v6c</summary>

-  Rebase

```
$ git range-diff gh/PASSWD_PROGRAM..gh/ep PASSWD_PROGRAM..ep 
1:  e3239be5addb ! 1:  d8cd8c5bc7bc configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
    @@ Commit message
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## configure.ac ##
    -@@ configure.ac: AC_SUBST([LIBSUBID_ABI_MINOR], [libsubid_abi_minor])
    +@@ configure.ac: AC_SUBST([LIBSUBID_ABI_MAJOR], [libsubid_abi_major])
    + AC_SUBST([LIBSUBID_ABI_MINOR], [libsubid_abi_minor])
      AC_SUBST([LIBSUBID_ABI_MICRO], [libsubid_abi_micro])
    - AC_SUBST([LIBSUBID_ABI], [libsubid_abi])
      
     -dnl Some hacks...
     -test "x$prefix" = "xNONE" && prefix="/usr"
2:  b33540fc7524 = 2:  4ab5f848fd39 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v6d</summary>

-  Rebase

```
$ git range-diff gh/PASSWD_PROGRAM..gh/ep PASSWD_PROGRAM..ep 
1:  d8cd8c5b = 1:  21a93750 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  4ab5f848 = 2:  bb50f101 src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v6e</summary>

-  Rebase

```
$ git range-diff gh/PASSWD_PROGRAM..gh/ep PASSWD_PROGRAM..ep 
1:  21a93750 = 1:  68432d56 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  bb50f101 = 2:  fe66955c src/Makefile.am: Use $() for consistency
```
</details>

<details>
<summary>v6f</summary>

-  Rebase

```
$ git range-diff gh/PASSWD_PROGRAM..gh/ep PASSWD_PROGRAM..ep 
1:  68432d56dd41 = 1:  fe06f2a62c51 configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables
2:  fe66955c0b88 = 2:  0ede07a3c148 src/Makefile.am: Use $() for consistency
```
</details>

alejandro-colomar committed:
configure.ac, src/Makefile.am: Follow GNU coding standards regarding directory variables

According to Sam, if we just remove the definitions, autotools will
provide the right values.

Note that this changes the default from /usr to /usr/local, which is the
right value according to the GNU conding standards (and also what the
BSDs prefer, according to one conversation I had with Ingo Schwarze from
OpenBSD).

Closes: <https://github.com/shadow-maint/shadow/issues/1229>
Link: <https://www.gnu.org/software/autoconf/manual/autoconf-2.72/html_node/Default-Prefix.html>
Reported-by: Chris Hofstaedtler <zeha@debian.org>
Cc: Sam James <sam@gentoo.org>
Signed-off-by: Alejandro Colomar <alx@kernel.org>

zeha commented:
#1227 relies on exec_prefix already being set in configure.ac; with both combined, i get this in config.h:
```
#define PASSWD_PROGRAM "NONE/bin/passwd"
```
but maybe you can just push a fix to #1227.

alejandro-colomar commented:
> #1227 relies on exec_prefix already being set in configure.ac; with both combined, i get this in config.h:
> 
> ```
> #define PASSWD_PROGRAM "NONE/bin/passwd"
> ```
> 
> but maybe you can just push a fix to #1227.

I bail out.  @thesamesam  please help.  :-)

thesamesam commented:
I'll look but probably tomorrow.

ikerexxe commented on src/Makefile.am:
```diff
@@ -2,8 +2,10 @@
 
-ubindir = ${prefix}/bin
-usbindir = ${prefix}/sbin
+prefix = /usr/local
+exec_prefix = ${prefix}
```
Shouldn't this be `/usr`?

ikerexxe replied:
Can you somehow indicate in the commit message that you have changed the default from `/usr` to `/usr/local`? This should warn developers and maintainers that the binary might not be installed where it was previously.

thesamesam commented on src/Makefile.am:
```diff
@@ -2,8 +2,10 @@
 
-ubindir = ${prefix}/bin
-usbindir = ${prefix}/sbin
+prefix = /usr/local
```
Are you sure this is right? I don't think we're supposed to set these at all. AFAIK, automake sets them for us.

thesamesam replied:
https://www.gnu.org/software/autoconf/manual/autoconf-2.72/html_node/Default-Prefix.html

alejandro-colomar replied:
No, I'm not sure at all.  I know those are the values we want, but I didn't know at all how they should be defined in autotools.  As they were being defined, I thought I had to define them somewhere.  I guess I'll just remove them.

Thanks!

alejandro-colomar replied:
I've removed it now; please check if it makes sense.

thesamesam reviewed (commented):
OK with that one change for `{}` vs `()`.

alejandro-colomar commented on src/Makefile.am:
```diff
@@ -24,7 +24,7 @@ AM_CFLAGS = $(LIBBSD_CFLAGS)
 # also /lib/libshadow.so.x.xx (if any) could be moved to /usr/lib
 # and installation would be much simpler (just two directories,
-# $prefix/bin and $prefix/sbin, no install-data hacks...)
+# $exec_prefix/bin and $exec_prefix/sbin, no install-data hacks...)
```
Sounds reasonable.  Feel free to send a patch.  :)

alejandro-colomar committed:
src/Makefile.am: Use $() for consistency

Some code was using $() and other ${}.  Use the same everywhere.

Reported-by: Sam James <sam@gentoo.org>
Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar commented on src/Makefile.am:
```diff
@@ -2,8 +2,8 @@
-ubindir = ${prefix}/bin
-usbindir = ${prefix}/sbin
+ubindir = ${exec_prefix}/bin
+usbindir = ${exec_prefix}/sbin
```
done

lslebodn commented on src/Makefile.am:
```diff
@@ -2,13 +2,13 @@
 	.indent.pro
 
-ubindir = ${prefix}/bin
-usbindir = ${prefix}/sbin
```
With the current state of MR, the same exec_prefix will be used for `ubindir` && `bindir` 
* https://www.gnu.org/software/automake/manual/html_node/Standard-Directory-Variables.html
So with the current state of MR  `ubindir` == `bindir` and  `usbindir` == `sbindir`.

Do we keep`ubindir` and `usbindir` in such case?
Or they are intentionally separate in Makefile.am so one can use make variables to have different  values for `ubindir` and `bindir` e.g.
```
make install ubindir=/tmp/bin usbindir=/tmp/sbin
```

Otherwise LGTM and +1 for removing hacks :-)

alejandro-colomar replied:
I want to remove them, but in a separate PR.  I'd say even in a separate release, just to make sure we don't break anything at some point.

alejandro-colomar replied:
Yes, I want that; however, I'd like the transition to be gradual, to make sure each patch is easy to review for correctness.

alejandro-colomar commented:
> > #1227 relies on exec_prefix already being set in configure.ac; with both combined, i get this in config.h:
> > ```
> > #define PASSWD_PROGRAM "NONE/bin/passwd"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but maybe you can just push a fix to #1227.
> 
> I bail out; too much autotools for me. @thesamesam please help. :-)

Has this been fixed?

ikerexxe commented:
As it is now, the patch does not break anything in the fedora packaging.

alejandro-colomar commented:
> > > #1227 relies on exec_prefix already being set in configure.ac; with both combined, i get this in config.h:
> > > ```
> > > #define PASSWD_PROGRAM "NONE/bin/passwd"
> > > ```
> > > 
> > > but maybe you can just push a fix to #1227.
> 
> Has this been fixed?
> 
> Cc: @zeha

Self answer: I still get this.

alejandro-colomar commented:
> > > > #1227 relies on exec_prefix already being set in configure.ac; with both combined, i get this in config.h:
> > > > ```
> > > > #define PASSWD_PROGRAM "NONE/bin/passwd"
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > but maybe you can just push a fix to #1227.
> > 
> > 
> > Has this been fixed?
> > Cc: @zeha
> 
> Self answer: I still get this.
> 
> @thesamesam , do you know why?

It's due to this line:
<https://github.com/shadow-maint/shadow/blob/b4beda846c4bd2c54d2b246ec18948acc47502c1/configure.ac#L106>

But is that line bad?  How should one use `$exec_prefix` within `configure.ac` properly?

Karlson2k commented:
Actually this need much deeper fix and hacks removals.
Currently the code is trying to abuse variables `prefix` and `exec_prefix` as "root" and "/usr" subdirs.
The code in https://github.com/shadow-maint/shadow/blob/d7ce7e868ed0dd6c17d4a422700e3da1e111e01c/src/Makefile.am#L5-L6 is trying to install to "/bin", "/usr/bin", "/sbin" and "/usr/sbin"
With global usr-merge it is much better to stop trying to differentiate "root" and "/usr".
If any downstream still needs to use different "root" and "/usr" it can be easily handled at packaging stage without any hacks.

For now, to fix this PR you just need to remove 
``` autoconf
AC_DEFINE_UNQUOTED([PASSWD_PROGRAM], ["$exec_prefix/bin/passwd"],
	[Path to passwd program.])
```
from configure.ac and put 
``` Makefile
DEFS += '-DPASSWD_PROGRAM="$(sbindir)/passwd"'
```
to https://github.com/shadow-maint/shadow/blob/d7ce7e868ed0dd6c17d4a422700e3da1e111e01c/lib/Makefile.am#L4

Note: empty assignment to "DEFS" in this file is wrong (or another hack?). It must be `DEFS +=`.

Alternatively (and less correct) you may just change configure.ac:
``` autoconf
AS_VAR_IF([exec_prefix],["NONE"],[:],
    [AC_DEFINE_UNQUOTED([PASSWD_PROGRAM], ["$exec_prefix/bin/passwd"],
	    [Path to passwd program.])]
)
```
The code is handling correctly undefined `PASSWD_PROGRAM` macro (with default pathname instead).

I will not be able to take my hands on it until February.

alejandro-colomar committed:
configure.ac, lib/Makefile.am: Respect $(bindir) in PASSWD_PROGRAM

Suggested-by: Evgeny Grin (Karlson2k) <k2k@drgrin.dev>
Signed-off-by: Alejandro Colomar <alx@kernel.org>

alejandro-colomar commented:
> You missed the proper quoting. You must use `'-DPASSWD_PROGRAM="$(sbindir)/passwd"'` or `-DPASSWD_PROGRAM=\"$(sbindir)/passwd\"` (but the latter may fail if whitespace is used in variable)

Ahh, sorry.  Why is this?  In regular makefiles, quoting is unnecessary.  Is this because of autotools?

Karlson2k commented:
At least some failures do not look like related.

alejandro-colomar committed:
lib/: Rename PASSWD_PROGRAM => PATH_PASSWD

For consistency with <paths.h> definitions.

Signed-off-by: Alejandro Colomar <alx@kernel.org>
